normandy/base/storage.py

```diff
@@ -9,6 +9,13 @@
 from storages.backends.gcloud import GoogleCloudStorage
 
 
+class RestrictedOverwriteStorageMixin(object):
+    def get_available_name(self, name, max_length=None):
+        if self.exists(name):
+            raise FileExistsError()
+        return super().get_available_name(name, max_length)
```

Comment:
This is still a TOCTOU error here. I think this is better than the way it was before, but it would be better if we could get an error thrown on over-write. That might just not be possible with Django storage engines though.

Reply:
Yeah I tried to find a solution that didn't suffer from the TOCTOU issue but there doesn't seem to be one. I'm open to ideas but this seems to be the way it's handled. I'm not sure if S3/GCP storage has options to disallow overwrite by folder or something?